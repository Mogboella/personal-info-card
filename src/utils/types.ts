export interface Name {
    firstName: string;
    lastName: string;
}

export interface Address {
    line1: string;
    line2?: string;
    line3?: string;
    town: string;
    county: string;
    eircode?: string;
    country?: string;
}

export interface Link {
    name: string;
    url: string;
}

export interface PhoneNumber {
    countryCode: string;
    number: string;
}

export interface Info {
    name: Name;
    email: string;
    links: Link[];
    number: PhoneNumber;
    address: Address;
}

export type CopyableField = keyof Info | `links.${number}`;