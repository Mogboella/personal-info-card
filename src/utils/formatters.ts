import { Address, Link, Name, PhoneNumber } from "./types";

/**
 * Formats a Name to FullName
 */
export function formatFullName(name: Name): string {
    return [name.firstName, name.lastName].filter(Boolean).join(' ')
}

/** Format Link */
export function formatLink(link: Link): string {
    return link.url
}

/** Formats Phone Number */
export function formatPhoneNumber(
    phone: PhoneNumber,
    options?: {
        format?: 'international' | 'national';
        includeType?: boolean;
        showExtension?: boolean;
    }
): string {
    const defaultOptions = {
        format: 'international',
        includeType: false,
        showExtension: true,
        ...options
    };

    // Clean all non-digit characters
    const cleanedNumber = phone.number.replace(/\D/g, '');
    const cleanedCountryCode = phone.countryCode.replace(/\D/g, '');

    let formattedNumber: string;

    switch (defaultOptions.format) {

        case 'national':
            formattedNumber = formatNationalNumber(cleanedNumber, cleanedCountryCode);
            break;

        case 'international':
        default:
            const internationalNumber = cleanedNumber.startsWith(cleanedCountryCode)
                ? cleanedNumber.slice(cleanedCountryCode.length)
                : cleanedNumber;
            formattedNumber = `+${cleanedCountryCode} ${formatNationalNumber(internationalNumber, cleanedCountryCode).replace(/^0/, '')}`;
    }

    return formattedNumber;
}

function formatNationalNumber(number: string, countryCode: string): string {
    switch (countryCode) {
        case '353': // Ireland
            // Remove international prefix if present
            const localNumber = number.startsWith('353')
                ? number.slice(3)
                : number;

            // Add leading 0 and format
            const withZero = localNumber.startsWith('0')
                ? localNumber
                : `0${localNumber}`;

            // Irish mobile (08x) or landline format
            if (/^08\d/.test(withZero)) {
                return `${withZero.slice(0, 3)} ${withZero.slice(3, 6)} ${withZero.slice(6)}`;
            }
            return `${withZero.slice(0, 2)} ${withZero.slice(2, 5)} ${withZero.slice(5)}`;

        case '1': // US/Canada
            return `(${number.slice(0, 3)}) ${number.slice(3, 6)}-${number.slice(6)}`;

        default:
            return number;
    }
}

/**
 * Formats an Irish address as a multi-line string
 */
export function formatAddressToString(address: Address, options?: {
    includeCountry?: boolean;
    eircodeUppercase?: boolean;
}): string {
    const lines = [
        address.line1,
        address.line2,
        address.line3,
        address.town,
        `Co. ${address.county}`,
        options?.eircodeUppercase ? address.eircode?.toUpperCase() : address.eircode,
        options?.includeCountry ? address.country || 'Ireland' : undefined
    ];

    return lines.filter(Boolean).join('\n');
}

/**
 * Single-line version for form autofill
 */
export function formatAddressSingleLine(address: Address): string {
    return [
        address.line1,
        address.line2,
        address.line3,
        address.town,
        `Co. ${address.county}`,
        address.eircode
    ].filter(Boolean).join(', ');
}