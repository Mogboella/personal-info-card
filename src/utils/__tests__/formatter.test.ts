import { formatAddressSingleLine, formatAddressToString, formatFullName, formatPhoneNumber } from "../formatters";
import { Address } from "../types";


// Name Format Testing 

describe('formatFullName', () => {
    it('combines first and last name', () => {
        expect(formatFullName({ firstName: 'John', lastName: 'Doe' })).toBe('John Doe');
    });

    it('handles empty last name', () => {
        expect(formatFullName({ firstName: 'John', lastName: '' })).toBe('John');
    });
});

// Phone Number Format Testing

describe('formatPhoneNumber', () => {
    describe('International format', () => {
        it('formats Irish numbers correctly', () => {
            expect(formatPhoneNumber({ countryCode: '353', number: '831234567' }))
                .toBe('+353 83 123 4567');
        });

        it('formats US numbers correctly', () => {
            expect(formatPhoneNumber({ countryCode: '1', number: '5551234567' }))
                .toBe('+1 (555) 123-4567');
        });
    });

    describe('National format', () => {
        it('formats Irish numbers without country code', () => {
            expect(formatPhoneNumber(
                { countryCode: '353', number: '831234567' },
                { format: 'national' }
            )).toBe('083 123 4567');
        });
    });
});

// Address Format Testing 

describe('address formatting', () => {
    const testAddress = {
        line1: '123 Main St',
        line2: 'Apt 4B',
        town: 'Dublin',
        county: 'Dublin',
        eircode: 'd01 abcd'
    };

    test('multi-line format', () => {
        expect(formatAddressToString(testAddress)).toMatchSnapshot();
    });

    test('single-line format', () => {
        expect(formatAddressSingleLine(testAddress))
            .toBe('123 Main St, Apt 4B, Dublin, Co. Dublin, d01 abcd');
    });

    test('uppercase eircode', () => {
        expect(formatAddressToString(testAddress, { eircodeUppercase: true }))
            .toContain('D01 ABCD');
    });
});

it('handles empty address line2', () => {
    const minimalAddress = {
        line1: '123 Main St',
        town: 'Dublin',
        county: 'Dublin'
    };
    expect(formatAddressSingleLine(minimalAddress as Address))
        .toBe('123 Main St, Dublin, Co. Dublin');
});