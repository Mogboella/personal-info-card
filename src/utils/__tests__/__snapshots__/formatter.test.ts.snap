// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`address formatting > multi-line format 1`] = `
"123 Main St
Apt 4B
Dublin
Co. Dublin
d01 abcd"
`;
